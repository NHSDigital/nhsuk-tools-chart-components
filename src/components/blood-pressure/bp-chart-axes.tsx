import {
  BpClassificationLevelType,
  BpClassificationLevel,
} from '@custom-types/blood-pressure.types';
import { BpBounds } from '@custom-types/classifications.types';
import { UpperBounds } from '@custom-types/threshold.types';
import React from 'react';
import './styles/bp-chart-axes.module.scss';

export enum AxisType {
  Systolic,
  Diastolic,
}

type BpGraphAxisLabelProps = Readonly<{
  axisType: AxisType;
  labelText: string;
  offsetPercent: number;
  applyLabelSpecificStyles?: boolean;
}>;

type BpGraphAxisProps = Readonly<{
  axisType: AxisType;
  upperBounds: UpperBounds<string>;
  lowerBound: number;
}>;

type BpGraphAxesProps = Readonly<{
  bounds: BpBounds;
}>;

export function BpGraphAxisValueLabel({
  axisType,
  labelText,
  offsetPercent,
  applyLabelSpecificStyles = true,
}: BpGraphAxisLabelProps): JSX.Element {
  const classPrefix = `${AxisType[axisType].toLowerCase()}-axis`;
  const genericMarkerClass = `nhsuk-bp-${classPrefix}-marker`;
  const specificMarkerClass = `nhsuk-bp-${classPrefix}-marker-${labelText}`;

  return (
    <div
      className={
        applyLabelSpecificStyles
          ? `${genericMarkerClass} ${specificMarkerClass}`
          : `${genericMarkerClass}`
      }
      style={
        axisType === AxisType.Systolic
          ? { bottom: `${offsetPercent}%` }
          : { left: `${offsetPercent}%` }
      }>
      <div className={`nhsuk-bp-${classPrefix}-label`}>{labelText}</div>
    </div>
  );
}

function boundaryPercentage(
  classificationUpperBoundValue: number,
  topUpperBoundValue: number,
  lowerBound: number
): number {
  return (
    ((classificationUpperBoundValue - lowerBound) /
      (topUpperBoundValue - lowerBound)) *
    100
  );
}

function classificationBoundaryPercentage(
  classification: BpClassificationLevelType,
  upperBounds: UpperBounds<string>,
  lowerBound: number
): number {
  return boundaryPercentage(
    upperBounds[classification].value!,
    upperBounds[BpClassificationLevel.High].value!,
    lowerBound
  );
}

function BpGraphAxis({
  axisType,
  upperBounds,
  lowerBound,
}: BpGraphAxisProps): JSX.Element {
  const extraDiastolicLabelConfig = {
    slightlyRaisedValueToCheck: 85,
    slightlyRaisedValueToAdd: 90,
  };
  return (
    <div>
      <BpGraphAxisValueLabel
        labelText={lowerBound.toString()}
        offsetPercent={0}
        axisType={axisType}
      />
      <BpGraphAxisValueLabel
        labelText={upperBounds[BpClassificationLevel.Low].value!.toString()}
        offsetPercent={classificationBoundaryPercentage(
          BpClassificationLevel.Low,
          upperBounds,
          lowerBound
        )}
        axisType={axisType}
      />
      <BpGraphAxisValueLabel
        labelText={upperBounds[BpClassificationLevel.Healthy].value!.toString()}
        offsetPercent={classificationBoundaryPercentage(
          BpClassificationLevel.Healthy,
          upperBounds,
          lowerBound
        )}
        axisType={axisType}
      />
      <BpGraphAxisValueLabel
        labelText={upperBounds[
          BpClassificationLevel.SlightlyRaised
        ].value!.toString()}
        offsetPercent={classificationBoundaryPercentage(
          BpClassificationLevel.SlightlyRaised,
          upperBounds,
          lowerBound
        )}
        axisType={axisType}
        applyLabelSpecificStyles={
          upperBounds[BpClassificationLevel.SlightlyRaised].value ===
          extraDiastolicLabelConfig.slightlyRaisedValueToCheck
        }
      />
      {axisType === AxisType.Diastolic &&
        upperBounds[BpClassificationLevel.SlightlyRaised].value ===
          extraDiastolicLabelConfig.slightlyRaisedValueToCheck && (
          <BpGraphAxisValueLabel
            labelText={extraDiastolicLabelConfig.slightlyRaisedValueToAdd.toString()}
            offsetPercent={boundaryPercentage(
              extraDiastolicLabelConfig.slightlyRaisedValueToAdd,
              upperBounds[BpClassificationLevel.High].value!,
              lowerBound
            )}
            axisType={axisType}
          />
        )}
      <BpGraphAxisValueLabel
        labelText={upperBounds[BpClassificationLevel.High].value!.toString()}
        offsetPercent={100}
        axisType={axisType}
      />
    </div>
  );
}

export function BpGraphAxes({ bounds }: BpGraphAxesProps): JSX.Element {
  return (
    <>
      <BpGraphAxis
        axisType={AxisType.Systolic}
        upperBounds={bounds.systolic.upper}
        lowerBound={bounds.systolic.lower}
      />
      <BpGraphAxis
        axisType={AxisType.Diastolic}
        upperBounds={bounds.diastolic.upper}
        lowerBound={bounds.diastolic.lower}
      />
    </>
  );
}
