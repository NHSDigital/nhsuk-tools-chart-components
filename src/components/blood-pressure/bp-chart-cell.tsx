import { BpCellConfiguration } from '@custom-types/blood-pressure.types';
import React from 'react';

import './styles/bp-chart-cell.module.scss';

export function BpChartCell({
  firstInRow = false,
  firstInColumn = false,
}: Readonly<BpCellConfiguration>): JSX.Element {
  let classNames = 'nhsuk-bp-graph-cell';

  if (firstInRow) {
    classNames += ` ${'nhsuk-bp-first-in-row'}`;
  }

  if (firstInColumn) {
    classNames += ` ${'nhsuk-bp-first-in-column'}`;
  }

  return <div className={classNames} />;
}
