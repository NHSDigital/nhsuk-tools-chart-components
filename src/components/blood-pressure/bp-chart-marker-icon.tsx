import React, { JSX } from 'react';

export function BpGraphMarkerIcon(): JSX.Element {
  return (
    <svg viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
      <g clipPath="url(#clip0_1346_102759)">
        <path
          d="M5.18718 8.36931L4.48008 9.07641L5.18718 9.78352L11.3744 15.9707L5.18718 22.1579L4.48008 22.865L5.18718 23.5721L8.36916 26.7541L9.07627 27.4612L9.78338 26.7541L15.9706 20.5669L22.1577 26.7541L22.8649 27.4612L23.572 26.7541L26.7539 23.5721L27.461 22.865L26.7539 22.1579L20.5668 15.9707L26.7539 9.78352L27.461 9.07641L26.7539 8.36931L23.572 5.18732L22.8649 4.48022L22.1577 5.18732L15.9706 11.3745L9.78338 5.18732L9.07627 4.48022L8.36916 5.18732L5.18718 8.36931Z"
          fill="black"
          stroke="white"
          strokeWidth="2"
        />
      </g>
    </svg>
  );
}
