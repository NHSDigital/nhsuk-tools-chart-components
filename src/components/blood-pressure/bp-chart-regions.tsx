import {
  BpClassificationLevel,
  BpClassificationLevelType,
} from '@custom-types/blood-pressure.types';
import { BpBounds } from '@custom-types/classifications.types';
import { UpperBounds } from '@custom-types/threshold.types';
import React from 'react';
import './styles/bp-chart-region-boundaries.module.scss';

export type BpGraphBoundaryProps = Readonly<{
  bounds: BpBounds;
  renderAsBoundary?: boolean;
}>;

function getClassificationBoundaryPercentage(
  lowerBound: number,
  upperBounds: UpperBounds<string>,
  classification: BpClassificationLevelType
): number {
  const normalisedBound = upperBounds[classification].value! - lowerBound;
  const normalisedUpperBound = upperBounds[3].value! - lowerBound;

  return (normalisedBound / normalisedUpperBound) * 100;
}

function getStyleForClassification(
  classification: BpClassificationLevelType
): string {
  return `nhsuk-bp-classification-${classification.toLowerCase()}`;
}

export function RegionsForClassification({
  classification,
  bounds,
  renderAsBoundary = false,
}: Readonly<
  BpGraphBoundaryProps & {
    classification: BpClassificationLevelType;
  }
>): JSX.Element {
  const width = getClassificationBoundaryPercentage(
    bounds.diastolic.lower,
    bounds.diastolic.upper,
    classification
  );
  const height = getClassificationBoundaryPercentage(
    bounds.systolic.lower,
    bounds.systolic.upper,
    classification
  );

  return (
    <div
      className={
        renderAsBoundary
          ? 'nhsuk-bp-region-boundary'
          : getStyleForClassification(classification)
      }
      style={{
        width: `${width}%`,
        height: `${height}%`,
      }}
    />
  );
}

export function BpGraphRegions({
  bounds,
  renderAsBoundary = false,
}: BpGraphBoundaryProps): JSX.Element {
  return (
    <>
      <RegionsForClassification
        classification={BpClassificationLevel.Low}
        {...{ bounds, renderAsBoundary }}
      />
      <RegionsForClassification
        classification={BpClassificationLevel.Healthy}
        {...{ bounds, renderAsBoundary }}
      />
      <RegionsForClassification
        classification={BpClassificationLevel.SlightlyRaised}
        {...{ bounds, renderAsBoundary }}
      />
      {!renderAsBoundary && (
        <RegionsForClassification
          classification={BpClassificationLevel.High}
          {...{ bounds, renderAsBoundary }}
        />
      )}
    </>
  );
}
