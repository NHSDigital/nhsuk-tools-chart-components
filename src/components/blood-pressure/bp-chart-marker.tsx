import Big from 'big.js';
import React, { JSX } from 'react';
import { BpGraphMarkerIcon } from './bp-chart-marker-icon';
import './styles/bp-chart-marker.module.scss';

// Changes rounding strategy to banker's rounding for unbias rounding
Big.RM = Big.roundHalfEven;

type BpGraphMarkerProps = Readonly<{
  systolicReading: number;
  diastolicReading: number;
  systolicLowerBound: number;
  systolicUpperBound: number;
  diastolicLowerBound: number;
  diastolicUpperBound: number;
}>;

export function BpGraphMarker({
  systolicReading,
  diastolicReading,
  systolicLowerBound,
  systolicUpperBound,
  diastolicLowerBound,
  diastolicUpperBound,
}: BpGraphMarkerProps): JSX.Element {
  const systolicDifference = systolicUpperBound - systolicLowerBound;
  const diastolicDifference = diastolicUpperBound - diastolicLowerBound;

  const bottomPercent = new Big(
    (systolicReading - systolicLowerBound) / systolicDifference
  ).times(100);
  const leftPercent = new Big(
    (diastolicReading - diastolicLowerBound) / diastolicDifference
  ).times(100);

  const roundingPrecision = 1;

  const roundedBottomPercent = bottomPercent.round(roundingPrecision);
  const roundedLeftPercent = leftPercent.round(roundingPrecision);

  return (
    <div
      className="nhsuk-bp-graph-marker"
      style={{
        left: `${roundedLeftPercent}%`,
        bottom: `${roundedBottomPercent}%`,
      }}>
      <BpGraphMarkerIcon />
    </div>
  );
}
