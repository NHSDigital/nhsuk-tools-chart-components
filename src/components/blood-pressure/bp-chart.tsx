import { BpBounds } from '@custom-types/classifications.types';
import {
  BpGraphLayout,
  BpReadingValue,
} from '@custom-types/blood-pressure.types';
import React from 'react';
import { BpGraphAxes } from './bp-chart-axes';
import { BpGraphMarker } from './bp-chart-marker';
import { BpGraphRegions } from './bp-chart-regions';
import { BpChartCell } from './bp-chart-cell';
import './styles/bp-chart.module.scss';

/**
 * The properties for @BpChart component
 *
 * @param {BpBounds} bounds - The blood pressure classification ranges
 * @param {BpGraphLayout} graphLayout - The graph layout contains rows configuration and systolic/diastolic chart labels
 * @param {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 */
export type BpChartProps = {
  bounds: BpBounds;
  graphLayout: BpGraphLayout;
  reading: BpReadingValue;
};

/**
 * A component to display the blood pressure chart
 *
 * @typedef {object} BpChartProps
 * @property {BpBounds} bounds - The blood pressure classification ranges
 * @property {BpGraphLayout} graphLayout - The graph layout contains rows configuration and systolic/diastolic chart labels
 * @property {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 *
 * @param {BpChartProps} props
 * @returns {JSX.Element}
 */
export function BpChart({
  bounds,
  graphLayout,
  reading,
}: Readonly<BpChartProps>): JSX.Element {
  const { systolic, diastolic } = reading;
  const { systolicLabel, diastolicLabel } = graphLayout;

  return (
    <div className="nhsuk-body-s">
      <div className="nhsuk-bp-systolic-legend-wrapper" aria-hidden="true">
        <div className="nhsuk-bp-systolic-legend-marker" />
        <span>{systolicLabel}</span>
      </div>

      <div className="nhsuk-bp-graph" aria-hidden="true">
        <BpGraphAxes bounds={bounds} />
        <BpGraphMarker
          systolicReading={systolic}
          diastolicReading={diastolic}
          systolicLowerBound={bounds.systolic.lower}
          systolicUpperBound={bounds.systolic.upper.high.value!}
          diastolicLowerBound={bounds.diastolic.lower}
          diastolicUpperBound={bounds.diastolic.upper.high.value!}
        />
        <BpGraphRegions bounds={bounds} />
        {graphLayout.rows.map((rowConfig) => (
          <div key={rowConfig.id} className="nhsuk-bp-graph-row">
            {rowConfig.cells.map((cellProps) => (
              <BpChartCell key={cellProps.id} {...cellProps} />
            ))}
          </div>
        ))}
        <BpGraphRegions bounds={bounds} renderAsBoundary />
      </div>

      <div className="nhsuk-bp-diastolic-legend-wrapper" aria-hidden="true">
        <div className="nhsuk-bp-diastolic-legend-marker" />
        <span>{diastolicLabel}</span>
      </div>
    </div>
  );
}
