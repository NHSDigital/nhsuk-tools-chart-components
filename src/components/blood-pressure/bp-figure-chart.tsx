import React, { JSX } from 'react';
import { ChartLegend, ChartLegendKey } from '@components/chart-legend';
import { BpBounds } from '@custom-types/classifications.types';
import {
  BpGraphLayout,
  BpLegendKeys,
  BpReadingValue,
} from '@custom-types/blood-pressure.types';
import { BpChart } from './bp-chart';
import { BpGraphMarkerIcon } from './bp-chart-marker-icon';
import './styles/bp-chart-panel.module.scss';

const populateBpChartLegendItems = (
  legendKeys: BpLegendKeys,
  legendMarkerText: string
): ChartLegendKey[] =>
  Object.entries(legendKeys)
    .map(
      ([classification, label]): ChartLegendKey => ({
        text: label,
        className: `nhsuk-bp-legend-key-classification-${classification}`,
      })
    )
    .concat({
      text: legendMarkerText,
      className: 'nhsuk-bp-legend-key-marker',
      element: <BpGraphMarkerIcon />,
    });

/**
 * The properties for @BpChartFigure component
 *
 * @param {BpBounds} bounds - The blood pressure classification ranges
 * @param {BpGraphLayout} graphLayout - The graph layout with row configuration and systolic/diastolic chart labels
 * @param {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 * @param {string }ariaLabel - The aria label for the chart figure
 * @param {BpLegendKeys} legendKeys - The legend keys for the chart figure
 * @param {string } legendMarkerText - The legend marker text
 */
export type BpChartFigureProps = {
  bounds: BpBounds;
  graphLayout: BpGraphLayout;
  reading: BpReadingValue;
  ariaLabel: string;
  legendKeys: BpLegendKeys;
  legendMarkerText: string;
};

/**
 * A component to display the blood pressure chart figure
 *
 * @typedef {object} BpChartFigureProps
 * @property {BpBounds} bounds - The blood pressure classification ranges
 * @property {BpGraphLayout} graphLayout - The graph layout with row configuration and systolic/diastolic chart labels
 * @property {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 * @property {string }ariaLabel - The aria label for the chart figure
 * @property {BpLegendKeys} legendKeys - The legend keys for the chart figure
 * @property {string } legendMarkerText - The legend marker text
 *
 * @param {BpChartFigureProps} props
 * @returns {JSX.Element}
 */
export function BpChartFigure({
  bounds,
  graphLayout,
  reading,
  ariaLabel,
  legendKeys,
  legendMarkerText,
}: Readonly<BpChartFigureProps>): JSX.Element {
  const bpLegendKeys = populateBpChartLegendItems(legendKeys, legendMarkerText);
  return (
    <figure className="nhsuk-u-margin-0">
      <figcaption className="nhsuk-u-visually-hidden">{ariaLabel}</figcaption>
      <BpChart bounds={bounds} reading={reading} graphLayout={graphLayout} />

      <ChartLegend legendKeys={bpLegendKeys} />
    </figure>
  );
}
