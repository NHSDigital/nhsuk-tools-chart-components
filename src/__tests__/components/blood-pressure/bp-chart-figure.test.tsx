import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import {
  BpGraphLayout,
  BpLegendKeys,
  BpReadingValue,
} from '@custom-types/blood-pressure.types';
import { BpBounds } from '@custom-types/classifications.types';
import {
  mockedBpClassificationBounds,
  mockedBpGraphLayout,
  mockedBptReading,
} from '@tests/__helpers__/mocks/bp-results';
import React from 'react';
import { BpChartFigure } from '@components/blood-pressure/bp-figure-chart';

const mockedAriaLabel = 'Blood Pressure figure description';

const mockedMarkerText = 'Your reading';
const mockedLegendKeys: BpLegendKeys = {
  high: 'High',
  slightlyraised: 'Slightly raised',
  healthy: 'healthy',
  low: 'low',
};

function BpChartFigureTestApp({
  classificationBounds = mockedBpClassificationBounds,
  reading = mockedBptReading,
  layout = mockedBpGraphLayout,
  ariaLabel = mockedAriaLabel,
  legendKeys = mockedLegendKeys,
  legendMarkerText = mockedMarkerText,
}: Readonly<{
  classificationBounds?: BpBounds;
  reading?: BpReadingValue;
  layout?: BpGraphLayout;
  ariaLabel?: string;
  legendKeys?: BpLegendKeys;
  legendMarkerText?: string;
}>): JSX.Element {
  return (
    <BpChartFigure
      bounds={classificationBounds}
      reading={reading}
      graphLayout={layout}
      ariaLabel={ariaLabel}
      legendKeys={legendKeys}
      legendMarkerText={legendMarkerText}
    />
  );
}

describe('The Blood pressure chart figure component', () => {
  it('Renders with the expected content', () => {
    const { container } = render(<BpChartFigureTestApp />);
    expect(container).toMatchSnapshot();
  });

  it('Renders a figcaption with aria label', () => {
    const { container } = render(
      <BpChartFigureTestApp ariaLabel={mockedAriaLabel} />
    );
    const figcaptionEl = container?.querySelector('figcaption');

    expect(figcaptionEl).not.toBeNull();
    expect(figcaptionEl?.classList).toContain('nhsuk-u-visually-hidden');
    expect(figcaptionEl?.innerHTML).toBe(mockedAriaLabel);
  });

  it('Renders with the expected marker keys', () => {
    const { container } = render(<BpChartFigureTestApp />);
    const legendCardEl = container?.querySelector('.nhsuk-chart-legend-card');

    expect(legendCardEl).not.toBeNull();

    const legendKeyWrappersEl = container?.querySelectorAll(
      '.nhsuk-chart-legend-key-wrapper'
    );
    expect(legendKeyWrappersEl).not.toBeNull();
    expect(legendKeyWrappersEl?.length).toBe(5);

    Object.entries(mockedLegendKeys).forEach(([key, value], index) => {
      const legendKeyWrapper = legendKeyWrappersEl![index];
      const legendKeyEl = legendKeyWrapper?.querySelector(
        `.nhsuk-bp-legend-key-classification-${key}`
      );
      expect(legendKeyEl).not.toBeNull();
      expect(legendKeyWrapper.innerHTML).toContain(value);
    });

    const legendKeyMarkerWrapperEl = legendKeyWrappersEl[4];
    const legendKeyMarkerEl = legendKeyMarkerWrapperEl?.querySelector(
      `.nhsuk-bp-legend-key-marker`
    );
    expect(legendKeyMarkerEl).not.toBeNull();
  });
});
