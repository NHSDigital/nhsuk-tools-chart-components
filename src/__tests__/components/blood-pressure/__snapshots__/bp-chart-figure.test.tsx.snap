// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Blood pressure chart figure component Renders with the expected content 1`] = `
<div>
  <figure
    class="nhsuk-u-margin-0"
  >
    <figcaption
      class="nhsuk-u-visually-hidden"
    >
      Blood Pressure figure description
    </figcaption>
    <div
      class="nhsuk-body-s"
    >
      <div
        aria-hidden="true"
        class="nhsuk-bp-systolic-legend-wrapper"
      >
        <div
          class="nhsuk-bp-systolic-legend-marker"
        />
        <span>
          Systolic (high number)
        </span>
      </div>
      <div
        aria-hidden="true"
        class="nhsuk-bp-graph"
      >
        <div>
          <div
            class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-70"
            style="bottom: 0%;"
          >
            <div
              class="nhsuk-bp-systolic-axis-label"
            >
              70
            </div>
          </div>
          <div
            class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-90"
            style="bottom: 18.181818181818183%;"
          >
            <div
              class="nhsuk-bp-systolic-axis-label"
            >
              90
            </div>
          </div>
          <div
            class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-120"
            style="bottom: 45.45454545454545%;"
          >
            <div
              class="nhsuk-bp-systolic-axis-label"
            >
              120
            </div>
          </div>
          <div
            class="nhsuk-bp-systolic-axis-marker"
            style="bottom: 63.63636363636363%;"
          >
            <div
              class="nhsuk-bp-systolic-axis-label"
            >
              140
            </div>
          </div>
          <div
            class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-180"
            style="bottom: 100%;"
          >
            <div
              class="nhsuk-bp-systolic-axis-label"
            >
              180
            </div>
          </div>
        </div>
        <div>
          <div
            class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-40"
            style="left: 0%;"
          >
            <div
              class="nhsuk-bp-diastolic-axis-label"
            >
              40
            </div>
          </div>
          <div
            class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-60"
            style="left: 25%;"
          >
            <div
              class="nhsuk-bp-diastolic-axis-label"
            >
              60
            </div>
          </div>
          <div
            class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-80"
            style="left: 50%;"
          >
            <div
              class="nhsuk-bp-diastolic-axis-label"
            >
              80
            </div>
          </div>
          <div
            class="nhsuk-bp-diastolic-axis-marker"
            style="left: 62.5%;"
          >
            <div
              class="nhsuk-bp-diastolic-axis-label"
            >
              90
            </div>
          </div>
          <div
            class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-120"
            style="left: 100%;"
          >
            <div
              class="nhsuk-bp-diastolic-axis-label"
            >
              120
            </div>
          </div>
        </div>
        <div
          class="nhsuk-bp-graph-marker"
          style="left: 50%; bottom: 63.6%;"
        >
          <svg
            fill="none"
            viewBox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_1346_102759)"
            >
              <path
                d="M5.18718 8.36931L4.48008 9.07641L5.18718 9.78352L11.3744 15.9707L5.18718 22.1579L4.48008 22.865L5.18718 23.5721L8.36916 26.7541L9.07627 27.4612L9.78338 26.7541L15.9706 20.5669L22.1577 26.7541L22.8649 27.4612L23.572 26.7541L26.7539 23.5721L27.461 22.865L26.7539 22.1579L20.5668 15.9707L26.7539 9.78352L27.461 9.07641L26.7539 8.36931L23.572 5.18732L22.8649 4.48022L22.1577 5.18732L15.9706 11.3745L9.78338 5.18732L9.07627 4.48022L8.36916 5.18732L5.18718 8.36931Z"
                fill="black"
                stroke="white"
                stroke-width="2"
              />
            </g>
          </svg>
        </div>
        <div
          class="nhsuk-bp-classification-low"
          style="width: 25%; height: 18.181818181818183%;"
        />
        <div
          class="nhsuk-bp-classification-healthy"
          style="width: 50%; height: 45.45454545454545%;"
        />
        <div
          class="nhsuk-bp-classification-slightlyraised"
          style="width: 62.5%; height: 63.63636363636363%;"
        />
        <div
          class="nhsuk-bp-classification-high"
          style="width: 100%; height: 100%;"
        />
        <div
          class="nhsuk-bp-graph-row"
        >
          <div
            class="nhsuk-bp-graph-cell nhsuk-bp-first-in-row nhsuk-bp-first-in-column"
          />
          <div
            class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
          />
          <div
            class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
          />
          <div
            class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
          />
          <div
            class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
          />
          <div
            class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
          />
        </div>
        <div
          class="nhsuk-bp-region-boundary"
          style="width: 25%; height: 18.181818181818183%;"
        />
        <div
          class="nhsuk-bp-region-boundary"
          style="width: 50%; height: 45.45454545454545%;"
        />
        <div
          class="nhsuk-bp-region-boundary"
          style="width: 62.5%; height: 63.63636363636363%;"
        />
      </div>
      <div
        aria-hidden="true"
        class="nhsuk-bp-diastolic-legend-wrapper"
      >
        <div
          class="nhsuk-bp-diastolic-legend-marker"
        />
        <span>
          Diastolic (low number)
        </span>
      </div>
    </div>
    <div
      aria-hidden="true"
    >
      <div
        class="nhsuk-card nhsuk-u-margin-top-6 nhsuk-u-margin-bottom-3 nhsuk-u-secondary-text-color nhsuk-body-s nhsuk-u-padding-3 nhsuk-chart-legend-card"
      >
        <div
          class="nhsuk-chart-legend-key-wrapper "
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-bp-legend-key-classification-high"
          />
          High
        </div>
        <div
          class="nhsuk-chart-legend-key-wrapper "
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-bp-legend-key-classification-slightlyraised"
          />
          Slightly raised
        </div>
        <div
          class="nhsuk-chart-legend-key-wrapper "
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-bp-legend-key-classification-healthy"
          />
          healthy
        </div>
        <div
          class="nhsuk-chart-legend-key-wrapper "
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-bp-legend-key-classification-low"
          />
          low
        </div>
        <div
          class="nhsuk-chart-legend-key-wrapper "
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-bp-legend-key-marker"
          >
            <svg
              fill="none"
              viewBox="0 0 32 32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                clip-path="url(#clip0_1346_102759)"
              >
                <path
                  d="M5.18718 8.36931L4.48008 9.07641L5.18718 9.78352L11.3744 15.9707L5.18718 22.1579L4.48008 22.865L5.18718 23.5721L8.36916 26.7541L9.07627 27.4612L9.78338 26.7541L15.9706 20.5669L22.1577 26.7541L22.8649 27.4612L23.572 26.7541L26.7539 23.5721L27.461 22.865L26.7539 22.1579L20.5668 15.9707L26.7539 9.78352L27.461 9.07641L26.7539 8.36931L23.572 5.18732L22.8649 4.48022L22.1577 5.18732L15.9706 11.3745L9.78338 5.18732L9.07627 4.48022L8.36916 5.18732L5.18718 8.36931Z"
                  fill="black"
                  stroke="white"
                  stroke-width="2"
                />
              </g>
            </svg>
          </div>
          Your reading
        </div>
      </div>
    </div>
  </figure>
</div>
`;
